File: 739_daily_temperatures/Solution.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        Stack<Pair<Integer, Integer>> nwd = new Stack<Pair<Integer, Integer>>();
        int[] daysToWait = new int[temperatures.length];
        
        for(int i = 0; i<temperatures.length; i++){
            while(!nwd.isEmpty() && nwd.peek().getKey() < temperatures[i]){
                daysToWait[nwd.peek().getValue()] = i - nwd.peek().getValue();
                nwd.pop();
            }
            nwd.push(new Pair<>(temperatures[i],i));
        }
        
        return daysToWait;
        
    }
}

File: 739_daily_temperatures/Solution.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        std::stack<pair<int,int>> nwd; //next warmer day
        std::vector<int> daysToWait(temperatures.size());
        std::fill(daysToWait.begin(), daysToWait.end(), 0);
        
        for(int i = 0; i<temperatures.size(); i++){
            while(!nwd.empty() && nwd.top().first < temperatures[i]){
                daysToWait[nwd.top().second] = i - nwd.top().second;
                nwd.pop();
            }
            nwd.push({temperatures[i], i});
        }
        
        return daysToWait;
    }
};
